### PR TITLE
examples: addition of example app

### PR DESCRIPTION
**Pull request by JavierDelgadoFernandez**
* Adds an example application.

Signed-off-by: Javier Delgado <javier.delgado.fernandez@cern.ch>

### CODE REVIEW

File: MANIFEST.in [hunk 1]
```diff
@@ -44,3 +44,4 @@ recursive-include tests *.py
 recursive-include invenio_collections *.po
 recursive-include invenio_collections *.mo
 recursive-include invenio_collections *.pot
+recursive-include examples *.xml
```
Comment (jirikuncar) [inline MANIFEST.in]: update copyright year

File: examples/app.py [hunk 2]
```diff
@@ -25,7 +25,23 @@
...
+
+    $ pip install dojson
+
+
+2. You will also need Invenio-Search in order to perform searches inside the
+   records.
+
+.. code-block:: console
+
+    $ pip install invenio-search==1.0.0a3
```
Comment (jirikuncar) [inline examples/app.py]: `>=` instead of `==`

File: examples/app.py [hunk 3]
```diff
@@ -35,35 +51,93 @@
...
+
+    $ flask -a app.py records create records.json
+
+
+5. Create a root collection (By default, the first collection is the root
+   node) using the CLI tool:
+
+.. code-block:: console
+
+    $ flask -a app.py collections create RootCollection
```
Comment (jirikuncar) [inline examples/app.py]: Demo or Examples?

File: examples/app.py [hunk 4]
```diff
@@ -35,35 +51,93 @@
...
+
+
+4. Load the file in the application importing the json file:
+
+.. code-block:: console
+
+    $ flask -a app.py records create records.json
+
+
+5. Create a root collection (By default, the first collection is the root
```
Comment (jirikuncar) [inline examples/app.py]: `Create a root collection using the CLI. This collection will become the root of new tree by default.`

--
please free to re-phrase it better

File: examples/app.py [hunk 5]
```diff
@@ -35,35 +51,93 @@
...
+      -q 'collections.primary:INSTITUTE'
+
+
+7. Create a sub-collection with a query to show only the nuclear research
+   center and its parent is `Institutes`:
+
+.. code-block:: console
+
+    $ flask -a app.py collections create Experimentelle -p Institutes \
+      -q 'citation_references_note.name_of_source:Experimentelle'
```
Comment (jirikuncar) [inline examples/app.py]: Experimentelle vs Experiments?

File: examples/app.py [hunk 6]
```diff
@@ -35,35 +51,93 @@
...
+from invenio_records import InvenioRecords
+from invenio_search import InvenioSearch, Query, current_search_client
 
 from invenio_collections import InvenioCollections
+from invenio_collections.models import Collection
 
 # Create Flask application
 app = Flask(__name__)
 app.config.update(
-    SQLALCHEMY_DATABASE_URI=os.getenv('SQLALCHEMY_DATABASE_URI',
```
Comment (jirikuncar) [inline examples/app.py]: please keep the `os.getenv` so people can easily change the database engine. Thanks

File: examples/app.py [hunk 7]
```diff
@@ -35,35 +51,93 @@
...
-Babel(app)
-Menu(app)
-Breadcrumbs(app)
 InvenioDB(app)
+InvenioRecords(app)
 InvenioCollections(app)
+InvenioSearch(app)
+
+
+@app.route('/<path:collection_name>/', methods=['GET'])
```
Comment (jirikuncar) [inline examples/app.py]: IMHO it should not be path but string.

File: examples/app.py [hunk 8]
```diff
@@ -35,35 +51,93 @@
...
+def index(collection_name):
+    """Query Elasticsearch using a collection"""
+    collection = Collection.query.filter_by(name=collection_name).one()
+    query = Query(collection.dbquery)
```
Comment (jirikuncar) [inline examples/app.py]: All records should have field `_collections` populated with the collection names. Please see tests made by @hachreak.

Comment (JavierDelgadoFernandez) [inline examples/app.py]: @jirikuncar What do you mean? I know there is a field with the `_collections` but my idea was to search with InvenioSearch a defined collection. Do you prefer to remove the InvenioSearch dependency and just iterate over all the records checking if they are in the specified collection?

Comment (jirikuncar) [inline examples/app.py]: If the dbquery is none then you should search in all children where dbquery is not none.

File: MANIFEST.in (hunk 1 above)
Comment (jirikuncar) [inline MANIFEST.in]: ping @JavierDelgadoFernandez

File: examples/app.py [hunk 9]
```diff
@@ -35,35 +51,96 @@
 
 You can find the database in `examples/app.db`.
 
-Run the development server:
+
+3. Transform the file `authority.xml` from MARC21 to a json file:
+
+.. code-block:: console
+
+    $ dojson do -i authority.xml -l marcxml marc21 > records.json
```
Comment (jirikuncar) [inline examples/app.py]: Please use the file from `invenio-records`.

--
see https://github.com/inveniosoftware/invenio/blob/master/scripts/populate-instance.sh#L115-L119

**Comment by JavierDelgadoFernandez** [comment]
@jirikuncar updated.

File: examples/app.py [hunk 10]
```diff
@@ -35,35 +51,121 @@
...
+
+    $ dojson do -i $demomarc21pathname -l marcxml marc21 > records.json
+
+
+5. Load the file in the application importing the json file:
 
 .. code-block:: console
 
-    $ flask -a app.py run -p 5000 -h '0.0.0.0'
+    $ flask -a app.py records create records.json
```
Comment (jirikuncar) [inline examples/app.py]: `$ flask -a app.py records create < records.json`

or 

`$ flask -a app.py records create -i records.json`

File: examples/app.py [hunk 11]
```diff
@@ -35,35 +51,121 @@
...
+
+
+6. Create a root collection using the CLI. This collection will become
+   the root of new tree by default:
+
+.. code-block:: console
+
+    $ flask -a app.py collections create Demo
```
Comment (jirikuncar) [inline examples/app.py]: If you create collections before uploading records you can search in `_collections` field in record metadata.

--
tip: make sure that `COLLECTIONS_REGISTER_RECORD_SIGNALS=True`

File: examples/app.py [hunk 12]
```diff
@@ -35,35 +51,121 @@
...
+            result.append(current_collection['node'].dbquery)
+        elif 'children' in current_collection:
+            result += get_collections_queries(current_collection['children'])
+    return result
+
+
+@app.route('/', methods=['GET'])
+def index():
+    """Query Elasticsearch using a collection"""
+    collection_name = request.values.get('collection')
```
Comment (jirikuncar) [inline examples/app.py]:
```python
collection_names = requests.values.getlist('collection')
# only if you would like to "validate" colection names
collections = Collection.query
if collection_names:
    collections = collections.filter(Collection.name.in_(collections))
assert len(collection_names) == collections.count()

response = current_search.client.search(
    body={'query': {'filtered': {'filter': {'terms': {'_collections': collection_names}}}}}
)
```

**Comment by JavierDelgadoFernandez** [comment]
@jirikuncar Rebased and updated.

**Pull request merged**